src/features/documentSymbolProvider.ts

```diff
@@ -4,20 +4,89 @@
  *--------------------------------------------------------------------------------------------*/
 
 import AbstractSupport from './abstractProvider';
+import * as protocol from '../omnisharp/protocol';
 import * as serverUtils from '../omnisharp/utils';
-import {toDocumentSymbol} from '../omnisharp/typeConvertion';
-import {DocumentSymbolProvider, SymbolInformation, TextDocument, CancellationToken} from 'vscode';
+import * as vscode from 'vscode';
 
-export default class OmnisharpDocumentSymbolProvider extends AbstractSupport implements DocumentSymbolProvider {
+import Structure = protocol.V2.Structure;
+import SymbolKinds = protocol.V2.SymbolKinds;
+import SymbolRangeNames = protocol.V2.SymbolRangeNames;
 
-    public async provideDocumentSymbols(document: TextDocument, token: CancellationToken): Promise<SymbolInformation[]> {
+export default class OmnisharpDocumentSymbolProvider extends AbstractSupport implements vscode.DocumentSymbolProvider {
 
-        return serverUtils.currentFileMembersAsTree(this._server, { FileName: document.fileName }, token).then(tree => {
-            let ret: SymbolInformation[] = [];
-            for (let node of tree.TopLevelTypeDefinitions) {
-                toDocumentSymbol(ret, node);
+    async provideDocumentSymbols(document: vscode.TextDocument, token: vscode.CancellationToken): Promise<vscode.SymbolInformation[]> {
+        const response = await serverUtils.codeStructure(this._server, { FileName: document.fileName }, token);
+        
+        if (response && response.Elements) {
+            return createSymbols(response.Elements, document.fileName);
+        }
+
+        return [];
+    }
+}
+
+function createSymbols(elements: Structure.CodeElement[], fileName: string): vscode.SymbolInformation[] {
+    let results: vscode.SymbolInformation[] = [];
+
+    function walkCodeElements(elements: Structure.CodeElement[], parentElementName?: string): void {
```

Comment:
Wonder if there's a helper we can extract...

Reply:
Yeah, I'll do that.